insns/generate.py

```diff
@@ -1101,6 +1146,150 @@ def insn_c_mvadd(insn, funct4, add, misa=MISA_C):
+        assign(f, "spec_rs1_addr", "insn_rs1")
+        assign(f, "spec_rd_addr", "insn_rd")
+        if wmode:
+            assign(f, "spec_rd_wdata", "spec_rd_addr ? {{`RISCV_FORMAL_XLEN-32{result[31]}}, result} : 0")
+        else:
+            assign(f, "spec_rd_wdata", "spec_rd_addr ? result : 0")
+        assign(f, "spec_pc_wdata", "rvfi_pc_rdata + 4")
+
+        footer(f)
+
+def insn_ext(insn, funct5, signed=False, bmode=False, misa=MISA_B):
+    with open("insn_%s.v" % insn, "w") as f:
+        header(f, insn)
+        format_b(f)
+        misa_check(f, misa)
+
+        if bmode:
+            result_width = "8"
+            opcode = "7'b 0011011"
+        else: # hmode
+            result_width = "16"
+            opcode = "7'b 0010011"
+
+        if signed:
+            funct7 = "0110000"
+            funct3 = "001"
+            opcode = "7'b 0010011"
+            result_extension = f"result[{result_width}-1]"
+        else:
+            funct7 = "0000100"
+            funct3 = "100"
+            opcode = "{3'b 011, `RISCV_FORMAL_XLEN != 32, 3'b 011}"
+            result_extension = "1'b 0"
+
+        print("", file=f)
+        print("  // %s instruction" % insn.upper(), file=f)
+        print("  wire [%s-1:0] result = rvfi_rs1_rdata[%s-1:0];" % (result_width, result_width), file=f)
+        assign(f, "spec_valid", "rvfi_valid && !insn_padding && insn_funct7 == 7'b %s && insn_funct5 == 5'b %s && insn_funct3 == 3'b %s && insn_opcode == %s" % (funct7, funct5, funct3, opcode))
+        assign(f, "spec_rs1_addr", "insn_rs1")
+        assign(f, "spec_rd_addr", "insn_rd")
+        assign(f, "spec_rd_wdata", "spec_rd_addr ? {{`RISCV_FORMAL_XLEN-%s{%s}}, result} : 0" % (result_width, result_extension))
+        assign(f, "spec_pc_wdata", "rvfi_pc_rdata + 4")
+
+        footer(f)
+
+def insn_bit(insn, funct6, funct3, expr, imode=False, misa=MISA_B):
+    with open("insn_%s.v" % insn, "w") as f:
+        header(f, insn)
+        format_rb(f)
+        misa_check(f, misa)
+
+        if imode:
+            opcode = "0010011"
+            xtra_shamt_check = "(!insn_shamt[5] || `RISCV_FORMAL_XLEN == 64)"
+            index = "insn_shamt & (`RISCV_FORMAL_XLEN - 1)"
+        else:
+            opcode = "0110011"
+            xtra_shamt_check = "!insn_shamt[5]"
+            index = "rvfi_rs2_rdata & (`RISCV_FORMAL_XLEN - 1)"
+
+        print("", file=f)
+        print("  // %s instruction" % insn.upper(), file=f)
+        print("  wire [`RISCV_FORMAL_XLEN-1:0] index = %s;" % index, file=f)
+        print("  wire [`RISCV_FORMAL_XLEN-1:0] result = %s;" % expr, file=f)
+        assign(f, "spec_valid", "rvfi_valid && !insn_padding && insn_funct6 == 6'b %s && insn_funct3 == 3'b %s && insn_opcode == 7'b %s && %s" % (funct6, funct3, opcode, xtra_shamt_check))
+        assign(f, "spec_rs1_addr", "insn_rs1")
+        assign(f, "spec_rs2_addr", "insn_rs2")
```

Comment:
This should be gated by `if not imode` as there is no meaningful rs2 address for immediate instructions